src/core/mod.rs

```diff
@@ -18,654 +18,665 @@
+	/// reply to arrive sooner if a mixnode misbehaves).
+	pub fn post_request(
+		&mut self,
+		destination: &mut Option<MixnodeId>,
+		data: &[u8],
+		num_surbs: usize,
+		ns: &dyn NetworkStatus,
+	) -> Result<Duration, PostErr> {
 		let mut rng = rand::thread_rng();
-		let mut result = Vec::new();
-		for _ in 0..count {
-			match self.topology.random_path_to(
-				&mut rng,
-				&start,
-				recipient,
-				num_hops.unwrap_or(self.num_hops),
-			) {
-				Some(path) => result.push(path),
-				None => return Err(Error::NoPath(Some(recipient.clone()))),
-			}
-		}
-		Ok(result)
-	}
 
-	fn random_cover_path(&self) -> Option<(MixPeerId, MixPublicKey)> {
-		// Select a random connected peer
-		let neighbors = self.neighbors();
+		// Split the message into fragments
+		let mut message_id = [0; MESSAGE_ID_SIZE];
+		rng.fill_bytes(&mut message_id);
+		let fragment_blueprints = match fragment_blueprints(&message_id, data, num_surbs) {
+			Some(fragment_blueprints)
+				if fragment_blueprints.len() <= self.config.max_fragments_per_message =>
+				fragment_blueprints,
+			_ => return Err(PostErr::TooManyFragments),
+		};
 
-		if neighbors.is_empty() {
-			return None
+		// Grab the session and check there's room in the queue
+		let session_index = destination.map_or_else(
+			|| {
+				self.session_status.phase.default_request_session() +
+					self.session_status.current_index
+			},
+			|destination| destination.session_index,
+		);
+		let session = post_session(&mut self.sessions, self.session_status, session_index)?;
+		// TODO Something better than this
+		if fragment_blueprints.len() > session.authored_packet_queue.remaining_capacity() {
+			return Err(PostErr::NotEnoughSpaceInQueue)
 		}
 
-		let mut rng = rand::thread_rng();
-		let n: usize = rng.gen_range(0..neighbors.len());
-		Some(neighbors[n].clone())
-	}
+		// Generate the packets and push them into the queue
+		let request_builder = RequestBuilder::new(
+			&mut rng,
+			&session.topology,
+			ns,
+			destination.map(|destination| destination.mixnode_index),
+		)
+		.map_err(PostErr::Topology)?;
+		let mut max_request_delay = Delay::zero();
+		let mut max_reply_delay = Delay::zero();
+		for fragment_blueprint in fragment_blueprints {
+			let (packet, delay) = request_builder
+				.build_packet(
+					&mut rng,
+					|fragment, rng| {
+						fragment_blueprint.write_except_surbs(fragment);
+						for surb in fragment_blueprint.surbs(fragment) {
+							// TODO Currently we don't clean up keystore entries on failure
+							let (id, keys) = self.surb_keystore.insert(rng, self.config.log_target);
+							let num_hops = self.config.num_hops;
+							let delay =
+								request_builder.build_surb(surb, keys, rng, &id, num_hops)?;
+							max_reply_delay = max(max_reply_delay, delay);
+						}
+						Ok(())
+					},
+					self.config.num_hops,
+				)
+				.map_err(PostErr::Topology)?;
+			session.authored_packet_queue.push(packet);
```

Comment:
Something I think of while reading this, would it make sense to only queue the first in authored_queue and apply a delay on the next ones and queue them in the forward queue, so we don't have this batch of request send at the same time?

Reply:
They're not really sent at the same time; request and reply packets are always dispatched in place of drop cover traffic, so they'll get dispatched over some period of time. The point of dispatching them in place of drop traffic is so that traffic analysis cannot reveal whether or not any messages were sent. It isn't quite perfect because reply packets are a bit different to drop cover packets -- their final hop will likely be to a non-mixnode. Maybe there is a good way to address this but I couldn't think of anything simple so I haven't done anything for now. In any case I definitely think we should not put request and reply packets in the forward queue, as that would result in extra traffic being generated when we are sending a request or reply, which could be noticeable.

Reply:
make sense, there is already a delay.
No solution either for the reply, does seems like a known limitation to me rather than something solvable (like first mixnet to send cover packet to the external node for some time, but sounds pretty costfull).